src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -287,6 +289,21 @@ private void HandleChannelException(IConnection connection, Exception ex)
             }
         }
 
+        private void HandlePreHandshakeException(Exception ex)
+        {
+            if (ex is MongoAuthenticationException)
+            {
+                _connectionPool.Clear();
```

Comment:
Now it's better, but actually we still process connection pool in two different places (here and in Invalidate method). Not sure how difficult to fully merge them?

Reply:
(T_T) :( 
I think unfortunately it'd be difficult to centralize the connection pool clearing logic since `Invalidate` is also called the post handshake exception handler with a potential value of `shouldClearConnectionPool: true.` (coming from `ShouldClearConnectionPoolForChannelException`. Since this ticket is blocking progress on other tickets, I'd prefer to keep it the way it is unless you strongly object :)